### PR TITLE
[kotlin] enforce fields in data classes

### LINKED ISSUE #14710: [BUG] [Kotlin] Data class must have at least one primary constructor parameter
#### Bug Report Checklist

- [ ] Have you provided a full/minimal spec to reproduce the issue?
- [ ] Have you validated the input using an OpenAPI validator ([example](https://apidevtools.org/swagger-parser/online/))?
- [ ] Have you [tested with the latest master](https://github.com/OpenAPITools/openapi-generator/wiki/FAQ#how-to-test-with-the-latest-master-of-openapi-generator) to confirm the issue still exists?
- [x] Have you searched for related issues/PRs?
- [ ] What's the actual output vs expected output?
- [ ] [Optional] Sponsorship to speed up the bug fix or feature request ([example](https://github.com/OpenAPITools/openapi-generator/issues/6178))

##### Description

I'm trying to generate a Kotlin class from a Swagger Spec.

Actual Kotlin Class
`class FooList : MutableList<Foo> by ArrayList()`

OpenAPI Generator output
`data class FooList () : kotlin.collections.ArrayList<Foo>()`

Build Error:
> Data class must have at least one primary constructor parameter

##### openapi-generator version

6.3.0

##### Generation Details

```kotlin
import org.openapitools.generator.gradle.plugin.tasks.GenerateTask

plugins {
  id("org.openapi.generator") version "6.3.0"
}

tasks.register<GenerateTask>() { ... }
```

##### Steps to reproduce

...

##### Suggest a fix

Tried adding [No-arg compiler plugin﻿](https://kotlinlang.org/docs/no-arg-plugin.html) but it would prob need to be ran at a certain time in the build?

```kotlin
plugins {
// id("org.jetbrains.kotlin.plugin.noarg") version "1.8.10"
  kotlin("plugin.noarg") version "1.7.21"
}

noArg {
  annotation("com.example.FooList")
}
```

### PR DESCRIPTION
**Pull request by jaqxues**
Fix #14710

### PR checklist
 
- [x] Read the [contribution guidelines](https://github.com/openapitools/openapi-generator/blob/master/CONTRIBUTING.md).
- [x] Pull Request title clearly describes the work in the pull request and Pull Request description provides details about how to validate the work. Missing information here may result in delayed response from the community.
- [x] Run the following to [build the project](https://github.com/OpenAPITools/openapi-generator#14---build-projects) and update samples:
  ```
  ./mvnw clean package 
  ./bin/generate-samples.sh ./bin/configs/*.yaml
  ./bin/utils/export_docs_generators.sh
  ``` 
  (For Windows users, please run the script in [Git BASH](https://gitforwindows.org/))
  Commit all changed files. 
  This is important, as CI jobs will verify _all_ generator outputs of your HEAD commit as it would merge with master. 
  These must match the expectations made by your contribution. 
  You may regenerate an individual generator by passing the relevant config(s) as an argument to the script, for example `./bin/generate-samples.sh bin/configs/java*`. 
  IMPORTANT: Do **NOT** purge/delete any folders/files (e.g. tests) when regenerating the samples as manually written tests may be removed.
- [x] File the PR against the [correct branch](https://github.com/OpenAPITools/openapi-generator/wiki/Git-Branches): `master` (upcoming 7.6.0 minor release - breaking changes with fallbacks), `8.0.x` (breaking changes without fallbacks)
- [x] If your PR is targeting a particular programming language, @mention the [technical committee](https://github.com/openapitools/openapi-generator/#62---openapi-generator-technical-committee) members, so they are more likely to review the pull request.

### CODE REVIEW

**Comment by jaqxues** [comment]
Can anyone of you review quickly? Thanks!
(Random subset of 3, not sure if I should've tagged less or more)

@dr4ke616 @stefankoppier @yutaka0m

**Comment by wing328** [comment]
can you please follow step 3 to update the samples so that the CI can test/verify the change?

cc @dr4ke616 (2018/08) @karismann (2019/03) @Zomzog (2019/04) @andrewemery (2019/10) @4brunu (2019/11) @yutaka0m (2020/03) @stefankoppier (2022/06)

**Comment by jaqxues** [comment]
@wing328 I ran the 3 commands when I made the commit already, and had no additional changes. Reran again now, git status remains clean.

Unless I should add additional things, I would say step 3 is handled.

**Comment by wing328** [comment]
you're right. samples are up-to-date

one thing i don't understand is how come there's no change in the samples as there are more test models with properties

i'll take another look later this week and run some tests

did you test the fix locally? I assume it works for your use cases

**Comment by jaqxues** [comment]
It does work. 

Out of interest, what are you expecting to change? If there were a sample about empty data, someone would probably have noticed that it generates code that does not compile. (I am new to this project and do not have any profound knowledge about the CI setup etc)

File: modules/openapi-generator/src/main/resources/kotlin-client/data_class.mustache [hunk 1]
```diff
@@ -74,7 +74,7 @@ import {{packageName}}.infrastructure.ITransformForStorage
 @Deprecated(message = "This schema is deprecated.")
 {{/isDeprecated}}
 {{>additionalModelTypeAnnotations}}
-{{#nonPublicApi}}internal {{/nonPublicApi}}{{#discriminator}}interface{{/discriminator}}{{^discriminator}}data class{{/discriminator}} {{classname}}{{^discriminator}} (
+{{#nonPublicApi}}internal {{/nonPublicApi}}{{#discriminator}}interface{{/discriminator}}{{^discriminator}}{{#hasVars}}data {{/hasVars}}class{{/discriminator}} {{classname}}{{^discriminator}} (
```
Comment (wing328) [inline modules/openapi-generator/src/main/resources/kotlin-client/data_class.mustache:77]: @jaqxues the `hasVars` is true when the model has at least one property.

what I don't understand is there's no change to the samples (e.g. kotlin petstore client) given that the test petstore pet has at least one model with properties.

Comment (jaqxues) [reply modules/openapi-generator/src/main/resources/kotlin-client/data_class.mustache:77]: This fixes an issue with models without properties, so the opposite case. A `data class` in kotlin is not allowed to have no properties

Comment (4brunu) [reply modules/openapi-generator/src/main/resources/kotlin-client/data_class.mustache:77]: Thanks for the clarification 👍

**Pull request merged**